RFC-0007-remote-functions.md

```diff
@@ -0,0 +1,219 @@
+
+
+## Background
+
+[prestodb/presto#14053](https://github.com/prestodb/presto/issues/14053) introduced the ability to define functions as being executed
+in a remote server, and made changes to the planner to accommodate remote function execution.  A limitation of the current implementation is that it
+presumes the functions returned from the function server is purely static.  However, it is common in many cloud data warehousing
+systems to allow for defining remote functions at runtime through `CREATE FUNCTION` statements.  This RFC proposes to extend the
+remote function design to allow for dynamic function registration and execution.
+
+Additionally, the current implementation of the remote function server plugin implementation is agnostic to the API of the function
+server.  The lack of a reference API, and additionally lack of documentation around function namespace managers, makes it
+challenging to create a remote function server: in addition to creating a new function server API, you also need to define a new
+namespace manager plugin.  To allow for a more consistent and unified experience, this RFC proposes a new REST API for remote function
+servers, which will allow for the definition of functions at runtime.  The hope is to reduce the work required to integrate a custom
+function server with Presto, firstly by reducing or eliminating the work required to write a new plugin, and secondly, by providing
+a reference implementation which itself will be extensible and hopefully cover most needs.
+
+### [Optional] Goals
+
+* Standardize on a single preferred API for remote function servers
+* Unify the metadata and execution of remote scalar functions under a single API
+* Allow for the definition of scalar functions at runtime
+
+### [Optional] Non-goals
+
+* Deprecate existing function server APIs
+* Provide support for aggregate functions or table-valued functions
+
+## Proposed Implementation
+
+### Design
+
+Fundamentally, the design of the REST API for remote functions will be based on the existing `FunctionNamespaceManager` interface SPI.
+Additionally, for Presto C++, currently there is no corresponding SPI for function namespace managers.  This RFC proposes to
+create a new REST-based implementation of the function execution framework.  By creating a REST API for remote functions that
+unifies the metadata and execution of remote functions, C++ deployments can similarly customize their function servers
+by implementing the REST API, in the same way that current Presto Java users can implement the `FunctionNamespaceManager` SPI.
+
+The REST API will power all `FunctionNamespaceManager` method implementations, including listing functions, retrieving function
+metadata, executing functions, and providing DDL support for functions (`ADD` and `DROP` support).
+
+#### Presto C++ special considerations
+
+The current Presto C++ implementation does not have a `FunctionNamespaceManager` SPI.  This RFC proposes to extend the Velox
```

Comment:
Are the function servers configured in a static list or is there a dynamic discovery of function servers ? Will the function server need to announce itself to a discovery service ?